File: src/main/java/random/telegramhomebot/controllers/HostController.java
package random.telegramhomebot.controllers;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import random.telegramhomebot.model.Host;
import random.telegramhomebot.repository.HostRepository;
import random.telegramhomebot.utils.pagination.PagerHelper;

import javax.annotation.Resource;
import javax.validation.Valid;
import java.util.Optional;
import java.util.UUID;

import static random.telegramhomebot.AppConstants.Hosts.ADD_EDIT_HOST_VIEW;
import static random.telegramhomebot.AppConstants.Hosts.DELETE_HOST_MAPPING;
import static random.telegramhomebot.AppConstants.Hosts.EDIT_HOST_BY_ID_MAPPING;
import static random.telegramhomebot.AppConstants.Hosts.EDIT_HOST_MAPPING;
import static random.telegramhomebot.AppConstants.Hosts.HOSTS_MAPPING;
import static random.telegramhomebot.AppConstants.Hosts.HOSTS_MODEL_ATTR;
import static random.telegramhomebot.AppConstants.Hosts.HOSTS_VIEW;
import static random.telegramhomebot.AppConstants.Hosts.HOST_ID_PATH_VAR;
import static random.telegramhomebot.AppConstants.Hosts.HOST_MAC_FIELD;
import static random.telegramhomebot.AppConstants.Hosts.HOST_MODEL_ATTR;
import static random.telegramhomebot.AppConstants.Hosts.REDIRECT_HOSTS;
import static random.telegramhomebot.AppConstants.Hosts.SAVE_HOST_MAPPING;
import static random.telegramhomebot.AppConstants.Messages.HOST_MAC_NOT_UNIQUE_MSG;

@Controller
@RequestMapping(HOSTS_MAPPING)
public class HostController {

	private static final int DEFAULT_CURRENT_PAGE = 0;

	@Resource
	private HostRepository hostRepository;

	@Value("${hosts.default.page.size}")
	private int defaultPageSize;
	@Value("${hosts.default.sorting}")
	private String defaultSorting;
	@Value("${hosts.default.sorting.direction}")
	private String defaultSortingDirection;

	@RequestMapping
	public String getAllHosts(@RequestParam("pageSize") Optional<Integer> pageSize,
							  @RequestParam("page") Optional<Integer> currentPage,
							  @RequestParam("sortBy") Optional<String> sortBy,
							  @RequestParam("direction") Optional<String> direction,
							  Model model) {

		PageRequest pageable = PagerHelper.getPageable(
				pageSize, defaultPageSize,
				currentPage, DEFAULT_CURRENT_PAGE,
				sortBy, defaultSorting,
				direction, defaultSortingDirection
		);

		Page<Host> hosts = hostRepository.findAll(pageable);
		model.addAttribute(HOSTS_MODEL_ATTR, hosts);

		PagerHelper.prepareModelForPager(
				model, pageable.getPageSize(), hosts.getTotalPages(), hosts.getNumber(), "hosts");
		return HOSTS_VIEW;
	}

	@RequestMapping(path = {EDIT_HOST_MAPPING, EDIT_HOST_BY_ID_MAPPING})
	public String editHostById(Model model, @PathVariable(HOST_ID_PATH_VAR) Optional<UUID> id) {
		model.addAttribute(HOST_MODEL_ATTR, Optional.ofNullable(id)
				.filter(Optional::isPresent)
				.flatMap(macOp -> hostRepository.findById(id.get()))
				.orElseGet(Host::new));
		return ADD_EDIT_HOST_VIEW;
	}

	@RequestMapping(path = DELETE_HOST_MAPPING)
	public String deleteHostById(@PathVariable(HOST_ID_PATH_VAR) UUID id) {
		hostRepository.deleteById(id);
		return REDIRECT_HOSTS;
	}

	@PostMapping(path = SAVE_HOST_MAPPING)
	public String createOrUpdateHost(@Valid Host host, BindingResult bindingResult) {
		if (bindingResult.hasErrors()) {
			return ADD_EDIT_HOST_VIEW;
		}
		Optional<Host> storedHost = hostRepository.findHostByMac(host.getMac());
		boolean saveNewHostWithExistingMac = storedHost.isPresent() && host.getId() == null;
		boolean editStoredHostMacToExisting = storedHost.isPresent() && !storedHost.get().getId().equals(host.getId());
		if (saveNewHostWithExistingMac || editStoredHostMacToExisting) {
			bindingResult.rejectValue(HOST_MAC_FIELD, HOST_MAC_NOT_UNIQUE_MSG);
			return ADD_EDIT_HOST_VIEW;
		}
		hostRepository.save(host);
		return REDIRECT_HOSTS;
	}
}

File: src/main/java/random/telegramhomebot/controllers/HostController.kt
package random.telegramhomebot.controllers

import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import random.telegramhomebot.AppConstants
import random.telegramhomebot.AppConstants.Hosts.ADD_EDIT_HOST_VIEW
import random.telegramhomebot.AppConstants.Hosts.DELETE_HOST_MAPPING
import random.telegramhomebot.AppConstants.Hosts.EDIT_HOST_BY_ID_MAPPING
import random.telegramhomebot.AppConstants.Hosts.EDIT_HOST_MAPPING
import random.telegramhomebot.AppConstants.Hosts.HOSTS_MAPPING
import random.telegramhomebot.AppConstants.Hosts.HOSTS_MODEL_ATTR
import random.telegramhomebot.AppConstants.Hosts.HOSTS_VIEW
import random.telegramhomebot.AppConstants.Hosts.HOST_ID_PATH_VAR
import random.telegramhomebot.AppConstants.Hosts.HOST_MAC_FIELD
import random.telegramhomebot.AppConstants.Hosts.HOST_MODEL_ATTR
import random.telegramhomebot.AppConstants.Hosts.REDIRECT_HOSTS
import random.telegramhomebot.AppConstants.Hosts.SAVE_HOST_MAPPING
import random.telegramhomebot.AppConstants.Redirects.ERROR_404_REDIRECT
import random.telegramhomebot.db.model.Host
import random.telegramhomebot.services.hosts.HostService
import random.telegramhomebot.utils.pagination.PagerHelper
import java.util.UUID
import javax.servlet.http.HttpServletRequest
import javax.validation.Valid

@Controller
@RequestMapping(HOSTS_MAPPING)
class HostController(private val hostService: HostService) {

    @Value("\${hosts.default.page.size}")
    private lateinit var defaultPageSize: Number

    @Value("\${hosts.default.sorting}")
    private lateinit var defaultSorting: String

    @Value("\${hosts.default.sorting.direction}")
    private lateinit var defaultSortingDirection: String

    @RequestMapping
    fun getAllHosts(
        @RequestParam("pageSize") pageSize: Int?,
        @RequestParam("page") currentPage: Int?,
        @RequestParam("sortBy") sortBy: String?,
        @RequestParam("direction") direction: String?,
        model: Model,
        request: HttpServletRequest?
    ): String {
        val pageSizeCookieName = "hostsPageSize"
        val pageable = PagerHelper.getPageable(
            pageSize ?: defaultPageSize.toInt(),
            currentPage ?: DEFAULT_CURRENT_PAGE,
            sortBy ?: defaultSorting,
            direction ?: defaultSortingDirection,
            pageSizeCookieName, request
        )
        val hosts = hostService.getAllHosts(pageable)
        model.addAttribute(HOSTS_MODEL_ATTR, hosts)
        PagerHelper.prepareModelForPager(
            model, hosts.totalPages, hosts.number, pageable.pageSize, pageSizeCookieName, HOSTS_MAPPING
        )
        return HOSTS_VIEW
    }

    @RequestMapping(path = [EDIT_HOST_MAPPING, EDIT_HOST_BY_ID_MAPPING])
    fun editHostById(model: Model, @PathVariable(HOST_ID_PATH_VAR) id: UUID?): String {
        val host =
            if (id == null) Host()
            else id.let { hostService.getHostById(it).orElse(null) } ?: return ERROR_404_REDIRECT

        model.addAttribute(HOST_MODEL_ATTR, host)
        return ADD_EDIT_HOST_VIEW
    }

    @RequestMapping(path = [DELETE_HOST_MAPPING])
    fun deleteHostById(@PathVariable(HOST_ID_PATH_VAR) id: UUID): String {
        hostService.deleteHostById(id)
        return REDIRECT_HOSTS
    }

    @PostMapping(path = [SAVE_HOST_MAPPING])
    fun createOrUpdateHost(@Valid host: Host, bindingResult: BindingResult): String {
        if (bindingResult.hasErrors()) {
            return ADD_EDIT_HOST_VIEW
        }
        val storedHost = hostService.getHostByMac(host.mac)
        val saveNewHostWithExistingMac = storedHost != null && host.id == null
        val editStoredHostMacToExisting = storedHost != null && storedHost.id != host.id
        if (saveNewHostWithExistingMac || editStoredHostMacToExisting) {
            bindingResult.rejectValue(HOST_MAC_FIELD, AppConstants.Messages.HOST_MAC_NOT_UNIQUE_MSG)
            return ADD_EDIT_HOST_VIEW
        }
        hostService.saveHost(host)
        return REDIRECT_HOSTS
    }

    companion object {
        private const val DEFAULT_CURRENT_PAGE = 1
    }
}
